Readability.js

```diff
@@ -831,6 +815,23 @@ Readability.prototype = {
     return node && node.nextElementSibling;
   },
 
+  // text similarity comparison
+  // 1 = same text, 0 = completely different text
+  // works the way that it splits both texts into words and then finds words that are unique in one or the other
+  // the result is given by the lower length of unique parts
+  _textSimilarity: function(textA, textB) {
+    var tokensA = textA.toLowerCase().split(this.REGEXPS.tokenize).filter(Boolean);
+    var tokensB = textB.toLowerCase().split(this.REGEXPS.tokenize).filter(Boolean);
+    if (tokensA.length === 0 || tokensB.length === 0) {
+      return 0;
+    }
+    var uniqTokensA = tokensA.filter(token => !tokensB.includes(token));
+    var uniqTokensB = tokensB.filter(token => !tokensA.includes(token));
+    var distanceA = uniqTokensA.join(" ").length / tokensA.join(" ").length;
+    var distanceB = uniqTokensB.join(" ").length / tokensB.join(" ").length;
+    return 1 - Math.min(distanceA, distanceB);
```

Comment:
So, as I understand it, if the detected article title is "Good food", and the header is "Good food: everything you need to know to cook like a pro", or whatever, this algorithm returns 1, because all of "Good food" occurs in the longer title, so one of `distanceA` and `distanceB` is 0, and so we return 1. And that would cause us to drop the h1. Correct? If so, I don't think that's quite optimal - can we avoid doing this if the detected title is the longer version of the same?

Reply:
yes that is exactly how it works now. I was inspired by the original H2 removal. I also tried using levensthein distance and that struggles in cases where it has the same words just in different order.

What if we do here `1 - Math.max(distanceA, distanceB)`. In that case the longer text will define the final number and in case as you described the text similarity will be low (~0.2) and so it will be kept.